### ISSUE TITLE
TypeError: cannot pickle 'weakref' object

### ISSUE THREAD

**Issue by Bhavay192** [Lightning-AI/lightning#18231]
### Bug description

I am facing this issue while using DDPSpawnPlugin of pytorchlightning for training on multiple GPUs.

Facing issue in version-1.5.1 but no option to select here.

Any help will be appreciated. Thanks

### What version are you seeing the problem on?

v1.6

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

**Comment by awaelchli** [comment]
Hi @Bhavay192 

You are not by any chance running into this? #15901
If yes, you find the solution here in my last comment on that issue: https://github.com/Lightning-AI/lightning/issues/15901#issuecomment-1505754665

If no, try also this:
- Use `strategy="ddp"` instead of "ddp_spawn"
- Upgrade Lightning to a more recent version

Without any code or any error trace, there is not much else I can do to help. Please consider providing this information, and also the version numbers of your packages, otherwise we won't be able to help or work on this. Thank you for your understanding.

**Comment by Bhavay192** [comment]
Hi @awaelchli, my PyTorch_lightning version is "1.5.1", PyTorch version = "1.10.0", 
This is my trainer config file - 
_target_: pytorch_lightning.Trainer

gpus: 0,1 # Specify GPU by CUDA_VISIBLE_DEVICES=0 
min_epochs: 1
max_epochs: 800
log_every_n_steps: 5
precision : 16
num_sanity_val_steps : 0 # This does not work with dp, only with ddp
check_val_every_n_epoch : 1
benchmark: True
overfit_batches: False
accelerator: ddp

I am initialising the DDPPlugin with this code,
if cfg.trainer.get('accelerator',None) == 'ddp': # for better performance in ddp mode
        print("HI")
        # plugs = DDPPlugin(find_unused_parameters=False)
        plugs = DDPSpawnPlugin(find_unused_parameters=False)
    else: 
        plugs = None
        
I'm initialising my trainer using this code 
trainer: Trainer = hydra.utils.instantiate(
            cfg.trainer, logger=logger, _convert_="partial", plugins=plugs ,callbacks=callbacks,
        )
        
Where can I use the strategy keyword. Also, I will update my version of both PyTorch and PyTorch_lightning and let u know

**Comment by awaelchli** [comment]
Yes sorry, this is quite an old version of Lightning and "strategy" as an argument doesn't exist there, it is still under the `plugin` argument. So use `plugs = DDPPlugin(find_unused_parameters=False)` in this case.

**Comment by Bhavay192** [comment]
I used this case only, and it gave me an error saying that some parameters are not taking part in loss calculation and needed to be searched, soo set the "find_unused_parameters" argument to True. I did that, and I faced with another issue. Should I reproduce the issue and send it? Will I face error with new version?

**Comment by awaelchli** [comment]
Yes, it is possible you need to set `find_unused_parameters=True` but that's unrelated. What I'm saying is try to use `plugs = DDPPlugin(...)` insteasd of `plugs = DDPSpawnPlugin(...)` and see if that helps.

> Should I reproduce the issue and send it? Will I face error with new version?

It's always best to have a small reproducible example to work with. Otherwise we can guess forever. Also, in your post you never sent any error message or stack trace, which would have been useful as well. 

> Will I face error with new version?

I can't say for sure, because you haven't provided much information for us to check that.

**Comment by Bhavay192** [comment]
Hi @awaelchli, I set the find_unused_parameters=False and the stack trace is this:
Error executing job with overrides: ['experiment=mDDPM/mDDPM']
Error executing job with overrides: ['experiment=mDDPM/mDDPM']
Traceback (most recent call last):
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 42, in <module>
    main()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 49, in decorated_main
    _run_hydra(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 367, in _run_hydra
    run_and_report(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 214, in run_and_report
    raise ex
Traceback (most recent call last):
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 368, in <lambda>
    lambda: hydra.run(
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 42, in <module>
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/hydra.py", line 110, in run
    _ = ret.return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 233, in return_value
    raise self._return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 160, in run_job
    ret.return_value = task_function(task_cfg)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 38, in main
    return train(config)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 44, in decorated_main
    return task_function(cfg_passthrough)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/src/train.py", line 136, in train
    trainer.fit(model, datamodule_train)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in fit
    self._call_and_handle_interrupt(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 682, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 770, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1193, in _run
    self._dispatch()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1272, in _dispatch
    self.training_type_plugin.start_training(self)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 202, in start_training
    self._results = trainer.run_stage()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1282, in run_stage
    return self._run_train()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1312, in _run_train
    self.fit_loop.run()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 234, in advance
    self.epoch_loop.run(data_fetcher)
    main()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 49, in decorated_main
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 195, in advance
    batch_output = self.batch_loop.run(batch, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py", line 88, in advance
    outputs = self.optimizer_loop.run(split_batch, optimizers, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 215, in advance
    result = self._run_optimization(
    _run_hydra(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 266, in _run_optimization
    self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 367, in _run_hydra
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 378, in _optimizer_step
    lightning_module.optimizer_step(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/lightning.py", line 1662, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/optimizer.py", line 164, in step
    trainer.accelerator.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 336, in optimizer_step
    self.precision_plugin.optimizer_step(model, optimizer, opt_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 163, in optimizer_step
    optimizer.step(closure=closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/optimizer.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/adam.py", line 92, in step
    loss = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 148, in _wrap_closure
    closure_result = closure()
    run_and_report(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 160, in __call__
    self._result = self.closure(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 214, in run_and_report
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 142, in closure
    step_output = self._step_fn()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 435, in _training_step
    training_step_output = self.trainer.accelerator.training_step(step_kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 216, in training_step
    return self.training_type_plugin.training_step(*step_kwargs.values())
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 439, in training_step
    return self.model(*args, **kwargs)
    raise ex
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 211, in run_and_report
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 873, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
making sure all `forward` function outputs participate in calculating loss.
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameter indices which did not receive grad for rank 0: 6
 In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error
    return func()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 368, in <lambda>
    lambda: hydra.run(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/hydra.py", line 110, in run
    _ = ret.return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 233, in return_value
    raise self._return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 160, in run_job
    ret.return_value = task_function(task_cfg)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 38, in main
    return train(config)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 44, in decorated_main
    return task_function(cfg_passthrough)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/src/train.py", line 136, in train
    trainer.fit(model, datamodule_train)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in fit
    self._call_and_handle_interrupt(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 682, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 770, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1193, in _run
    self._dispatch()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1272, in _dispatch
    self.training_type_plugin.start_training(self)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 202, in start_training
    self._results = trainer.run_stage()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1282, in run_stage
    return self._run_train()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1312, in _run_train
    self.fit_loop.run()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 234, in advance
    self.epoch_loop.run(data_fetcher)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 195, in advance
    batch_output = self.batch_loop.run(batch, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py", line 88, in advance
    outputs = self.optimizer_loop.run(split_batch, optimizers, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 215, in advance
    result = self._run_optimization(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 266, in _run_optimization
    self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 378, in _optimizer_step
    lightning_module.optimizer_step(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/lightning.py", line 1662, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/optimizer.py", line 164, in step
    trainer.accelerator.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 336, in optimizer_step
    self.precision_plugin.optimizer_step(model, optimizer, opt_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 163, in optimizer_step
    optimizer.step(closure=closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/optimizer.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/adam.py", line 92, in step
    loss = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 148, in _wrap_closure
    closure_result = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 160, in __call__
    self._result = self.closure(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 142, in closure
    step_output = self._step_fn()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 435, in _training_step
    training_step_output = self.trainer.accelerator.training_step(step_kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 216, in training_step
    return self.training_type_plugin.training_step(*step_kwargs.values())
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/ddp.py", line 439, in training_step
    return self.model(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1102, in _call_impl
    return forward_call(*input, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/nn/parallel/distributed.py", line 873, in forward
    if torch.is_grad_enabled() and self.reducer._rebuild_buckets():
RuntimeError: Expected to have finished reduction in the prior iteration before starting a new one. This error indicates that your module has parameters that were not used in producing loss. You can enable unused parameter detection by passing the keyword argument `find_unused_parameters=True` to `torch.nn.parallel.DistributedDataParallel`, and by
making sure all `forward` function outputs participate in calculating loss.
If you already have done the above, then the distributed data parallel module wasn't able to locate the output tensors in the return value of your module's `forward` function. Please include the loss function and the structure of the return value of `forward` of your module when reporting this issue (e.g. list, dict, iterable).
Parameter indices which did not receive grad for rank 1: 6
 In addition, you can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print out information about which particular parameters did not receive gradient on this rank as part of this error

**Comment by Bhavay192** [comment]
However, when I set the find_unused_parameters=True, this error comes up:


Error executing job with overrides: ['experiment=mDDPM/mDDPM']
Traceback (most recent call last):
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 42, in <module>
    main()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 49, in decorated_main
    _run_hydra(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 367, in _run_hydra
    run_and_report(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 214, in run_and_report
    raise ex
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 368, in <lambda>
    lambda: hydra.run(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/hydra.py", line 110, in run
    _ = ret.return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 233, in return_value
    raise self._return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 160, in run_job
    ret.return_value = task_function(task_cfg)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 38, in main
    return train(config)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 44, in decorated_main
    return task_function(cfg_passthrough)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/src/train.py", line 136, in train
    trainer.fit(model, datamodule_train)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in fit
    self._call_and_handle_interrupt(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 682, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 770, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1193, in _run
    self._dispatch()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1272, in _dispatch
    self.training_type_plugin.start_training(self)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 202, in start_training
    self._results = trainer.run_stage()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1282, in run_stage
    return self._run_train()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1312, in _run_train
    self.fit_loop.run()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 234, in advance
    self.epoch_loop.run(data_fetcher)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 195, in advance
    batch_output = self.batch_loop.run(batch, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py", line 88, in advance
    outputs = self.optimizer_loop.run(split_batch, optimizers, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 215, in advance
    result = self._run_optimization(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 266, in _run_optimization
    self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 378, in _optimizer_step
    lightning_module.optimizer_step(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/lightning.py", line 1662, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/optimizer.py", line 164, in step
    trainer.accelerator.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 336, in optimizer_step
    self.precision_plugin.optimizer_step(model, optimizer, opt_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 163, in optimizer_step
    optimizer.step(closure=closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/optimizer.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/adam.py", line 92, in step
    loss = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 148, in _wrap_closure
    closure_result = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 160, in __call__
    self._result = self.closure(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 155, in closure
    self._backward_fn(step_output.closure_loss)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 327, in backward_fn
    self.trainer.accelerator.backward(loss, optimizer, opt_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 311, in backward
    self.precision_plugin.backward(self.lightning_module, closure_loss, *args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 91, in backward
    model.backward(closure_loss, optimizer, *args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/lightning.py", line 1444, in backward
    loss.backward(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/_tensor.py", line 307, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/__init__.py", line 154, in backward
    Variable._execution_engine.run_backward(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/function.py", line 199, in apply
    return user_fn(self, *args)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/utils/checkpoint.py", line 138, in backward
    torch.autograd.backward(outputs_with_grad, args_with_grad)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/__init__.py", line 154, in backward
    Variable._execution_engine.run_backward(
RuntimeError: Expected to mark a variable ready only once. This error is caused by one of the following reasons: 1) Use of a module parameter outside the `forward` function. Please make sure model parameters are not shared across multiple concurrent forward-backward passes. or try to use _set_static_graph() as a workaround if this module graph does not change during training loop.2) Reused parameters in multiple reentrant backward passes. For example, if you use multiple `checkpoint` functions to wrap the same part of your model, it would result in the same set of parameters been used by different reentrant backward passes multiple times, and hence marking a variable ready multiple times. DDP does not support such use cases in default. You can try to use _set_static_graph() as a workaround if your module graph does not change over iterations.
Parameter at index 284 has been marked as ready twice. This means that multiple autograd engine  hooks have fired for this particular parameter during this iteration. You can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print parameter names for further debugging.
Error executing job with overrides: ['experiment=mDDPM/mDDPM']
Traceback (most recent call last):
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 42, in <module>
    main()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 49, in decorated_main
    _run_hydra(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 367, in _run_hydra
    run_and_report(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 214, in run_and_report
    raise ex
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 211, in run_and_report
    return func()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/utils.py", line 368, in <lambda>
    lambda: hydra.run(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/_internal/hydra.py", line 110, in run
    _ = ret.return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 233, in return_value
    raise self._return_value
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/core/utils.py", line 160, in run_job
    ret.return_value = task_function(task_cfg)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/run.py", line 38, in main
    return train(config)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/hydra/main.py", line 44, in decorated_main
    return task_function(cfg_passthrough)
  File "/home/ukhalid/hasan_work/Bhavay_work/mDDPM_new/src/train.py", line 136, in train
    trainer.fit(model, datamodule_train)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 735, in fit
    self._call_and_handle_interrupt(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 682, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 770, in _fit_impl
    self._run(model, ckpt_path=ckpt_path)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1193, in _run
    self._dispatch()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1272, in _dispatch
    self.training_type_plugin.start_training(self)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 202, in start_training
    self._results = trainer.run_stage()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1282, in run_stage
    return self._run_train()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 1312, in _run_train
    self.fit_loop.run()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/fit_loop.py", line 234, in advance
    self.epoch_loop.run(data_fetcher)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 195, in advance
    batch_output = self.batch_loop.run(batch, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/batch/training_batch_loop.py", line 88, in advance
    outputs = self.optimizer_loop.run(split_batch, optimizers, batch_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 215, in advance
    result = self._run_optimization(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 266, in _run_optimization
    self._optimizer_step(optimizer, opt_idx, batch_idx, closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 378, in _optimizer_step
    lightning_module.optimizer_step(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/lightning.py", line 1662, in optimizer_step
    optimizer.step(closure=optimizer_closure)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/optimizer.py", line 164, in step
    trainer.accelerator.optimizer_step(self._optimizer, self._optimizer_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 336, in optimizer_step
    self.precision_plugin.optimizer_step(model, optimizer, opt_idx, closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 163, in optimizer_step
    optimizer.step(closure=closure, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/optimizer.py", line 88, in wrapper
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 28, in decorate_context
    return func(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/optim/adam.py", line 92, in step
    loss = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 148, in _wrap_closure
    closure_result = closure()
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 160, in __call__
    self._result = self.closure(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 155, in closure
    self._backward_fn(step_output.closure_loss)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/loops/optimization/optimizer_loop.py", line 327, in backward_fn
    self.trainer.accelerator.backward(loss, optimizer, opt_idx)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/accelerators/accelerator.py", line 311, in backward
    self.precision_plugin.backward(self.lightning_module, closure_loss, *args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/plugins/precision/precision_plugin.py", line 91, in backward
    model.backward(closure_loss, optimizer, *args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/pytorch_lightning/core/lightning.py", line 1444, in backward
    loss.backward(*args, **kwargs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/_tensor.py", line 307, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/__init__.py", line 154, in backward
    Variable._execution_engine.run_backward(
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/function.py", line 199, in apply
    return user_fn(self, *args)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/utils/checkpoint.py", line 138, in backward
    torch.autograd.backward(outputs_with_grad, args_with_grad)
  File "/home/ukhalid/.conda/envs/Bhavay1/lib/python3.9/site-packages/torch/autograd/__init__.py", line 154, in backward
    Variable._execution_engine.run_backward(
RuntimeError: Expected to mark a variable ready only once. This error is caused by one of the following reasons: 1) Use of a module parameter outside the `forward` function. Please make sure model parameters are not shared across multiple concurrent forward-backward passes. or try to use _set_static_graph() as a workaround if this module graph does not change during training loop.2) Reused parameters in multiple reentrant backward passes. For example, if you use multiple `checkpoint` functions to wrap the same part of your model, it would result in the same set of parameters been used by different reentrant backward passes multiple times, and hence marking a variable ready multiple times. DDP does not support such use cases in default. You can try to use _set_static_graph() as a workaround if your module graph does not change over iterations.
Parameter at index 284 has been marked as ready twice. This means that multiple autograd engine  hooks have fired for this particular parameter during this iteration. You can set the environment variable TORCH_DISTRIBUTED_DEBUG to either INFO or DETAIL to print parameter names for further debugging.

**Comment by Bhavay192** [comment]
I'm unable to find the solution for both :(

**Comment by Bhavay192** [comment]
Hi @awaelchli , can you help please? I am facing some deadline that's why.

**Comment by Bhavay192** [comment]
Hi @awaelchli , any update on this please? Thanks

**Comment by awaelchli** [comment]
Hey @Bhavay192 I'm sorry but it is difficult for me to help here. I see a couple possibilities:

a) You're using an old version of Lightning (1.6) and in the new version this problem was fixed. Solution: Upgrade Lightning
b) The problem is in any of the Lighntning version. Solution: We need a reproducible code sample to be able to know which conditions lead to this. 
c) The problem is not in Lightning, and instead in the user (your) code. Solution: Based on the error, you need to check a few things. 1) Check that you are not calling backward yourself or calling it twice in a row for some reason. 2) Have you overridden the `backward()` or `optimizer_step()` hooks in the LightningModule? If yes, something might be wrong there. d) Run your script with `TORCH_DISTRIBUTED_DEBUG=DETAIL python train.py ...` as suggested in the error message to see which parameter name shows up.  
e) Replace the model with a dummy model that's very simple, like a single linear layer. If this works -> Issue is with the model. If this doesn't work -> Issue is might be with the training code.
f) If there is gradient checkpointing in your model, you'll probably have to turn it off (as suggested in the error message). 

These are the main points that come to mind. Please understand that it is almost impossible to help you with this little information you have provided, and that the Lightning version you are using is quite old for me to remember any details or known issues from back then.

**Comment by Bhavay192** [comment]
Hi @awaelchli, my transformer model is using gradient checkpointing and that is causing the error I believe. I used gradient checkpointing to reduce the memory usage of my transformer model. Any suggestions how can I reduce the memory usage and also train the model on multiple GPUs as my dataset size is also quite huge and I need to train on multiple GPUs for faster training.

**Comment by awaelchli** [comment]
How big is the model (number of parameters) and how much gpu memory do you have available?
